### PR TITLE
OMSAgent Multi-Homing Support

### PR DESCRIPTION
**Pull request by robbiezhang**
Support to onboard the agent to multiple workspaces
Register omsagent-<workspace id> daemon per workspace
Add more command in omsadmin.sh for multiple workspace operation
Update OMS heartbeat to collect workspace-specific process usage.
Manage the ports of Syslog and Monitor agent so that ports won't conflict for different workspaces
Enable the LogRotate per workspace
Implement the update script for migrating the files and directories from old agent to new.
@Microsoft/omsagent-devs

### CODE REVIEW

File: installer/scripts/omsadmin.sh [hunk 1]
```diff
@@ -424,11 +449,28 @@ onboard()
...
+    configure_syslog
 
+    configure_monitor_agent
+
+    configure_logrotate
+
+    # If a test is not in progress then register omsagent as a service and start the agent 
+    if [ -z "$TEST_WORKSPACE_ID" -a -z "$TEST_SHARED_KEY" ]; then
+        $SERVICE_CONTROL start $WORKSPACE_ID 
+    fi
```
Comment (sugr4) [inline installer/scripts/omsadmin.sh]: This closing fi needs to be removed so that omsconfig and omsConsistencyInvoker are set up when tests are not run.

Comment (lagalbra) [inline installer/scripts/omsadmin.sh]: +1 The indentation seems to agree in GitHub's UI.

File: installer/scripts/omsadmin.sh [hunk 2]
```diff
@@ -424,11 +449,28 @@ onboard()
...
+        echo $MULTI_HOMING_MARKER > $CONF_DIR/.multihoming_marker
+    fi
+
+    configure_syslog
 
+    configure_monitor_agent
+
+    configure_logrotate
+
+    # If a test is not in progress then register omsagent as a service and start the agent
```
Comment (lagalbra) [inline installer/scripts/omsadmin.sh]: I think Suju's comment from commit 467e845 should be used here.

**Comment by NarineM** [comment]
Can you point me to test cases that onboard to multiple workspaces ? Example:
1. Onboarding succeeded for both workspaces
2. Onboarding failed for one workspace , but succeeded with other
3. OMSAgent installed but not onboarded, then onboarded using conf to just one workspace.

**Comment by NarineM** [comment]
I could not see the documentation updated for end  user with multi -homing how to's and examples. Can you point it in case I missed them?

**Comment by robbiezhang** [comment]
Here is the system test for onboarding. https://github.com/Microsoft/OMS-Agent-for-Linux/blob/FB-multihoming/test/installer/scripts/omsadmin_systest.rb
I've added a test case to onboard 2 workspaces, list, remove, etc.

**Comment by robbiezhang** [comment]
I'm still working on the onboarding docs update. It will be in https://github.com/Microsoft/OMS-Agent-for-Linux/blob/FB-multihoming/docs/OMS-Agent-for-Linux.md

File: installer/datafiles/base_omsagent.data [hunk 3]
```diff
@@ -140,63 +145,24 @@ WriteInstallInfo
...
     else
         cp /etc/opt/microsoft/omsagent/sysconf/omsagent.d/oms.conf /etc/collectd/collectd.conf.d/oms.conf
     fi
-    cp /etc/opt/microsoft/omsagent/sysconf/omsagent.d/collectd.conf ${{CONF_DIR}}/omsagent.d/collectd.conf
+    #cp /etc/opt/microsoft/omsagent/sysconf/omsagent.d/collectd.conf ${{CONF_DIR}}/omsagent.d/collectd.conf
     rm /etc/collectd_marker.conf
 fi
 
-# The real conf directory has to be owned by omsagent to be readable by the omsagent daemon
-chown -R omsagent:omsagent ${{CONF_DIR}}
```
Comment (robbiezhang) [inline installer/datafiles/base_omsagent.data]: This script is run in the installation phase. It will create the /etc/opt/microsoft/omsagent/conf folder and grant permission to omsagent. We now move the logic to the onboarding phase, and change the folder to /etc/opt/microsoft/omsagent/<workspace id>/conf

File: docs/OMS-Agent-for-Linux.md [hunk 4]
```diff
@@ -186,6 +204,57 @@ SHARED_KEY=<Shared Key>
...
+cd /opt/microsoft/omsagent/bin
+sudo ./omsadmin.sh -w <WorkspaceID> -s <Shared Key> -m <Multi-Homing Marker>
+```
+NOTE: Secondary workspace is currently unable to pull the configuration from OMS service. We are working on it.
+
+## Onboard a secondary workspace using a file
+Reference [Onboarding using a file](#onboarding-using-a-file)
+Add the following line into /etc/omsagent-onboard.conf
+```
+MULTI_HOMING_MARKER=<Any marker you choose>
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: What is an example of a marker you could choose? I'm not sure you provide an example in this doc.

File: docs/OMS-Agent-for-Linux.md [hunk 5]
```diff
@@ -186,6 +204,57 @@ SHARED_KEY=<Shared Key>
...
+```
+
+# Manage Workspaces
+From 1.3.0-1, OMSAgent supports onboarding to multiple workspaces. Here are the commands for workspace management
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: end this line with a colon or period:
... commands for workspace management:

File: docs/OMS-Agent-for-Linux.md [hunk 6]
```diff
@@ -186,6 +204,57 @@ SHARED_KEY=<Shared Key>
...
+
+## List all workspaces
+```
+sudo sh /opt/microsoft/omsagent/bin/omsadmin.sh -l
+```
+
+Sample result for an agent onboarded to 2 workspaces:
+```
+Primary Workspace: 660c7bcd-28d2-453a-84bd-8523e396f600    Success(OMSAgent Registered)
+Workspace: 08d4b0c0-7fac-4159-987c-000271282eff    Success(OMSAgent Registered)
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: Have we provided real workspace IDs on GitHub before? I think it might be better to give mocked workspace IDs here, like abcdefghi-1234-5678-9101-jklmnopqrst

File: docs/OMS-Agent-for-Linux.md [hunk 7]
```diff
@@ -186,6 +204,57 @@ SHARED_KEY=<Shared Key>
...
+```
+
+## Remove a workspace
+```
+sudo sh /opt/microsoft/omsagent/bin/omsadmin.sh -x <Workspace ID>
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: capitalization again; use "workspace id"

File: docs/OMS-Agent-for-Linux.md [hunk 8]
```diff
@@ -186,6 +204,57 @@ SHARED_KEY=<Shared Key>
...
+```
+
+## Remove all workspaces
+```
+sudo sh /opt/microsoft/omsagent/bin/omsadmin.sh -X
+```
+
+# Manage omsagent Daemon
+From 1.3.0-1, we will register omsagent daemon for each onboarded workspace. The daemon name is omsagent-<Workspace-ID>
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: add a qualifier here: "register one omsagent daemon"

Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: capitalization with daemon name: workspace id

File: docs/OMS-Agent-for-Linux.md [hunk 9]
```diff
@@ -186,6 +205,57 @@ SHARED_KEY=<Shared Key>
...
+You can use /opt/microsoft/omsagent/bin/service_control command to operate the daemon.
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: end this line with a colon rather than a period, since it refers to the command below

File: docs/OMS-Agent-for-Linux.md [hunk 10]
```diff
@@ -186,6 +205,57 @@ SHARED_KEY=<Shared Key>
...
+
+```
+sudo sh /opt/microsoft/omsagent/bin/service_control start|stop|restart|enable|disable [<Workspace ID>]
+```
+
+The workspace id is an optional parameter. If it is specified, it will only operate on the workspace specific daemon.
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: nit: include a hyphen: workspace-specific

File: docs/OMS-Agent-for-Linux.md [hunk 11]
```diff
@@ -186,6 +205,57 @@ SHARED_KEY=<Shared Key>
...
+Otherwise, it will operates on all the daemons.
```
Comment (lagalbra) [inline docs/OMS-Agent-for-Linux.md]: Change to "it will operate on all the daemons"

File: installer/datafiles/base_omsagent.data [hunk 12]
```diff
@@ -72,7 +76,6 @@ MAINTAINER:              'Microsoft Corporation'
 /opt/microsoft/omsagent/plugin/filter_security_baseline.rb;             source/code/plugins/filter_security_baseline.rb;       744; root; root
 /opt/microsoft/omsagent/plugin/security_baseline_lib.rb;                source/code/plugins/security_baseline_lib.rb;          744; root; root
 /opt/microsoft/omsagent/plugin/omsbaseline;                             source/code/plugins/omsbaseline_${{PFARCH}};           755; root; root
-/etc/opt/microsoft/omsagent/sysconf/omsagent.d/oms_audits.xml;          installer/conf/omsagent.d/oms_audits.xml;              644; root; root
```
Comment (lagalbra) [inline installer/datafiles/base_omsagent.data]: Why was this line removed?

File: docs/OMS-Agent-for-Linux.md (hunk 4 above)
Comment (robbiezhang) [inline docs/OMS-Agent-for-Linux.md]: Any string will be good. It will be shown in the list workspace command

File: docs/OMS-Agent-for-Linux.md (hunk 6 above)
Comment (robbiezhang) [inline docs/OMS-Agent-for-Linux.md]: workspace id is not a secret. I can mock them

File: installer/datafiles/base_omsagent.data (hunk 12 above)
Comment (robbiezhang) [inline installer/datafiles/base_omsagent.data]: It's moved up to the section with all the files placed in /etc/opt/... folder
